### PR TITLE
Add admin page for HA tracker information.

### PR DESCRIPTION
**Pull request by cstyan**
It wasn't clear to me what the right way to serve the page was, and whether for other pages (ex: the ring pages) if we serve the page from every distributor/ingester or not.

Signed-off-by: Callum Styan <callumstyan@gmail.com>

### CODE REVIEW

**Comment by cstyan** [comment]
cc @tomwilkie @gouthamve @jtlisi

File: pkg/cortex/modules.go [hunk 1]
```diff
@@ -170,6 +170,7 @@ func (t *Cortex) initDistributor(cfg *Config) (err error) {
 
 	t.server.HTTP.HandleFunc("/all_user_stats", t.distributor.AllUserStatsHandler)
 	t.server.HTTP.Handle("/api/prom/push", t.httpAuthMiddleware.Wrap(http.HandlerFunc(t.distributor.PushHandler)))
+	t.server.HTTP.Handle("/ha-tracker", t.distributor.Replicas)
```
Comment (tomwilkie) [inline pkg/cortex/modules.go]: `Replicas` is nil is HA tracking is disabled, so this will panic.

Comment (tomwilkie) [reply pkg/cortex/modules.go]: In general, having this be nil when HA tracking is disabled is a pain.  I would always create the struct and internalise the enabled state within that struct.

File: pkg/distributor/ha_tracker_http.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+<html>
+	<head>
+		<meta charset="UTF-8">
+		<title>Cortex HA Tracker Status</title>
+	</head>
+	<body>
+		<h1>Cortex HA Tracker Status</h1>
+		<p>Current time: {{ .Now }}</p>
+		<form action="" method="POST">
+			<input type="hidden" name="csrf_token" value="$__CSRF_TOKEN_PLACEHOLDER__">
```
Comment (tomwilkie) [inline pkg/distributor/ha_tracker_http.go]: This is only needed if the page has an "action" - ie a button.  Same with the form.

File: pkg/distributor/ha_tracker_http.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			<table width="100%" border="1">
+				<thead>
+					<tr>
+						<th>User ID</th>
+						<th>Cluster</th>
+						<th>Replica</th>
+						<th>Timestamp</th>
```
Comment (tomwilkie) [inline pkg/distributor/ha_tracker_http.go]: When the lease runs out or when we last took the lease?  Both would be good.

File: pkg/distributor/ha_tracker_http.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+					</tr>
+				</thead>
+				<tbody>
+					{{ range .Elected }}
+					<tr>
+						<td>{{ .UserID }}</td>
+						<td>{{ .Cluster }}</td>
+						<td>{{ .Replica }}</td>
+						<td>{{ .Timestamp }}</td>
```
Comment (tomwilkie) [inline pkg/distributor/ha_tracker_http.go]: Does this render in a useable form?

**Comment by tomwilkie** [comment]
Looks really good - couple minor nits, but hopefully merge tomorrow.

File: pkg/distributor/distributor.go [hunk 5]
```diff
@@ -166,13 +164,11 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
...
-		replicas, err := newClusterTracker(cfg.HATrackerConfig)
-		if err != nil {
-			return nil, err
-		}
-		d.replicas = replicas
+	replicas, err := newClusterTracker(cfg.HATrackerConfig)
+	if err != nil {
+		return nil, err
 	}
+	d.Replicas = replicas
```
Comment (tomwilkie) [inline pkg/distributor/distributor.go]: Nit: I prefer if this goes above the building of the Distributor now.

File: pkg/distributor/distributor.go [hunk 6]
```diff
@@ -109,7 +109,6 @@ type Config struct {
 	BillingConfig billing.Config             `yaml:"billing,omitempty"`
 	PoolConfig    ingester_client.PoolConfig `yaml:"pool,omitempty"`
 
-	EnableHATracker bool            `yaml:"enable_ha_tracker,omitempty"`
```
Comment (tomwilkie) [inline pkg/distributor/distributor.go]: Not sure you can remove this completely - the distributor won't start if you don't provide consul/etc address for the ha tracker.

The idea was to move this flag into the HA tracker config instead, and have that struct 'know' if it were enabled or not - and not start the consul client, loop etc if it is.

File: pkg/distributor/ha_tracker_http.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+		electedReplicas = append(electedReplicas, struct {
+			UserID, Cluster, Replica            string
+			ElectedAt, UpdateTime, FailoverTime time.Time
+		}{
+			UserID:       chunks[0],
+			Cluster:      chunks[1],
+			Replica:      desc.Replica,
+			ElectedAt:    timestamp.Time(desc.ReceivedAt),
+			UpdateTime:   timestamp.Time(desc.ReceivedAt).Add(h.cfg.UpdateTimeout),
+			FailoverTime: timestamp.Time(desc.ReceivedAt).Add(h.cfg.FailoverTimeout),
```
Comment (tomwilkie) [inline pkg/distributor/ha_tracker_http.go]: Might be worth having these as durations, so they print out a bit more concisely?

Are UpdateTimeout and FailoverTimeout configurable per user?

**Comment by tomwilkie** [comment]
Couple more nits.

File: pkg/distributor/distributor.go (hunk 6 above)
Comment (cstyan) [reply pkg/distributor/distributor.go]: :man_facepalming:

File: pkg/distributor/ha_tracker_http.go (hunk 7 above)
Comment (cstyan) [reply pkg/distributor/ha_tracker_http.go]: Update/FailoverTimeout aren't configurable per user. Whether we accept HA samples from a user and what cluster/replica flags to look for are.

Comment (cstyan) [reply pkg/distributor/ha_tracker_http.go]: Durations via what call? `time.Until()`?

**Comment by cstyan** [comment]
@tomwilkie PTAL. In the most recent commit I've reverted the removal of/moved the `EnableHATracker` cfg flag to the HA tracker cfg from the Distributor cfg, added a prefix flag to the HA tracker flag registration (so we don't have to manually prefix each of it's flags with `distributor.`), and moved any checks for `if EnableHATracker` from distributor.go to ha_tracker.go. 

This means the HA Tracker struct is always created, but it only creates the client to it's KV store if `EnableHATracker == true`. Also if there's a misconfiguration of `EnableHATracker == false` but `acceptHASamples == true` for a user ID, we'll still accept samples that contain both HA tracking labels (cluster/replica), the HA tracker `checkReplica` will tell the distributor to accept any sample if `EnableHATracker == false`.

File: pkg/distributor/ha_tracker_http.go (hunk 7 above)
Comment (tomwilkie) [reply pkg/distributor/ha_tracker_http.go]: Yeah, `Until`.

File: pkg/distributor/ha_tracker.go [hunk 8]
```diff
@@ -80,27 +80,32 @@ type haTracker struct {
...
-	FailoverTimeout time.Duration
+	FailoverTimeout time.Duration `yaml:"ha_tracker_failover_timeout"`
 
 	KVStore kv.Config
 }
 
-// RegisterFlags adds the flags required to config this to the given FlagSet
-func (cfg *HATrackerConfig) RegisterFlags(f *flag.FlagSet) {
+// RegisterFlags adds the flags required to config this to the given FlagSet. If prefix is not an empty string it should end with a period.
+func (cfg *HATrackerConfig) RegisterFlags(prefix string, f *flag.FlagSet) {
```
Comment (tomwilkie) [inline pkg/distributor/ha_tracker.go]: TBH not sure the prefix is the right way to go here - its not like we're planning on using the HA tracker anywhere else, are we?

Comment (cstyan) [reply pkg/distributor/ha_tracker.go]: Not that I can think of, but it makes the flag registration a bit cleaner and less chance of new flags missing the `distributor.` prefix imo. I can revert this if you disagree.

File: pkg/distributor/distributor.go (hunk 5 above)
Comment (cstyan) [reply pkg/distributor/distributor.go]: I'm assuming you mean move the creation of the HA tracker to above the creation of the distributor, but leave this assignment. I've done that in the latest commit.

**Comment by cstyan** [comment]
test failures look like a flake?

**Comment by bboreham** [comment]
`TestChunkStoreRandom` failures noted at #1522

It _could_ be a real problem, but not one introduced by this PR.

File: pkg/distributor/ha_tracker.go (hunk 8 above)
Comment (tomwilkie) [reply pkg/distributor/ha_tracker.go]: We have to be really careful changing flag names.  For instance you dropped the dot between distributor and ha-tracker.  This has the potential to break rollouts and piss off users.

File: pkg/distributor/distributor.go [hunk 9]
```diff
@@ -139,7 +139,7 @@ type Config struct {
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	cfg.BillingConfig.RegisterFlags(f)
 	cfg.PoolConfig.RegisterFlags(f)
-	cfg.HATrackerConfig.RegisterFlags(f)
+	cfg.HATrackerConfig.RegisterFlags("distributor.", f)
```
Comment (cstyan) [inline pkg/distributor/distributor.go]: @tomwilkie I'm confused, I had the prefix as `distributor.` but your commit removing the prefix shows it was just `distributor`

File: pkg/distributor/ha_tracker.go (hunk 8 above)
Comment (cstyan) [reply pkg/distributor/ha_tracker.go]: https://github.com/cortexproject/cortex/pull/1546/commits/8ffe4b9064485e5cc293892209bdb391c8257e9a#r312024987

File: pkg/distributor/distributor.go (hunk 9 above)
Comment (tomwilkie) [reply pkg/distributor/distributor.go]: `¯\_(ツ)_/¯`

File: pkg/distributor/ha_tracker_http.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+var trackerTmpl *template.Template
+
+func init() {
+	trackerTmpl = template.Must(template.New("ha-tracker").Parse(trackerTpl))
+}
+
+func (h *haTracker) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	h.electedLock.RLock()
+	defer h.electedLock.RUnlock()
```
Comment (gouthamve) [inline pkg/distributor/ha_tracker_http.go]: Can we unlock right after we use it? I see we hold this lock until the network transfer is over and I see it as a potential DOS target, with someone deliberately holding the connection, as holding the lock will mean requests fail.

**Pull request merged**